### ISSUE TITLE
The arc-tanh formula is wrong

### ISSUE THREAD

**Issue by ChinhQT** [woheller69/Arity#10]
Hi,
I don't know why this one is wrong. The Arity's result is tanh^-1(2%^9) = 5.5511151E-16
But the correct result is tanh^-1(2%^9) = 5.12E-16.

**Comment by woheller69** [comment]
are you sure?  2%^9 equals 5.12E-16
Should atanh of this value be the same?

**Comment by ChinhQT** [comment]
tanh and atanh (2%^9) = 5.12E-16. They have the same result. I've just tested on other app. It still has the same result. You can check them again.

**Comment by woheller69** [comment]
I checked the formula and it is correct.
So I guess these are precision issues of Java's Math library.

**Comment by Fontan030** [comment]
Yes, looks like Java's library has some general issues with trigonometry. For example, sin(11π) gives 4.899E-15, sin(2000π) is evaluated as -6.428E-13 (while both expressions obviously should be equal to 0)

**Comment by woheller69** [comment]
Java double seems to be good for about 15 digits...

**Comment by ChinhQT** [comment]
Are you sure ? I don't think it's an error of Java's library. The ClevCal on Android, it works well. The result is correct. I don't why.

**Comment by woheller69** [comment]
btw, Java has already problems subtracting 56.3 - 55.7. It gives 0.59999999
I recently fixed that by using BigDecimal for subtractions

**Comment by ChinhQT** [comment]
With the formula 0.5 * Math.log(1. + x) / (1. - x)) 
x = 2%^9. The result is wrong. 
When you use x = 2%^8. The Arity works well. The result is correct. I don't know why.

**Comment by woheller69** [comment]
well I guess it is the 1 + 2%^9 or 1 - 2%^9. One or both of them probably exceeds the possibilty of "double"

**Comment by ChinhQT** [comment]
Do you have any suggestion to resolve this issue, please ?

**Comment by woheller69** [comment]
No. For high-precision stuff use another calculator.

**Comment by woheller69** [comment]
next version will have a better check for pi multiples, so sin(11PI) and sin(2000PI) should be 0
But that is all I can do.

**Comment by supsm** [comment]
> well I guess it is the 1 + 2%^9 or 1 - 2%^9. One or both of them probably exceeds the possibilty of "double"

Yep, I looked into this and it's actually a known issue.  
libc (llvm) gets around this by [using a taylor series approximation for small values](https://github.com/llvm/llvm-project/blob/main/libc/src/math/generic/atanhf.cpp#L48), the code boils down to 1/11 x^11 + 1/9 x^9 + 1/7 x^7 + 1/5 x^5 + 1/3 x^3 + x. glibc (gnu) [uses `log1p`](https://github.com/bminor/glibc/blob/master/sysdeps/ieee754/flt-32/e_atanhf.c#L61), which calculates ln(1 + x) with greater precision around x=0.  
Neither clang (llvm), gcc (gnu), nor msvc (microsoft) optimized `0.5 * log(1 + (x + x) / (1 - x))` to `0.5 * log1p((x + x) / (1 - x))`, so I doubt java would as well.  

If java has `log1p`, it should be an easy replacement for small input values, and if not, the taylor approximation also works.

**Comment by woheller69** [comment]
I am using my own fork https://github.com/woheller69/ArityEngine
Will apply the fix there :-)
Maybe we should add a test case also

**Comment by woheller69** [comment]
if 0.5 * log1p((x + x) / (1 - x)) should be used:

Why are you proposing

     0.5 * Math.log1p(twox + twox * x / (1 - x));

Shouldn't it be simply

     0.5 * Math.log1p(twox / (1 - x));

**Comment by supsm** [comment]
They're mathematically the same thing. I'm not *exactly* sure why the gcc wizards chose this, but it must have been purposeful somehow as they had the other option for another case. Besides, they're some of the world's smartest people working on critical infrastructure, so there's plenty of reason for it to be as optimized as possible.  

I think it might be because for exceedingly small values of x, `twox + twox * x / (1 - x)` guarantees at least `twox` (`x + x` is just an increase in floating point exponent with no effect on precision, and operator precedence dictates that the addition is done last; `twox * x / (1 - x)` can never be negative), which is an alright approximation at those values. `twox / (1 - x)`, on the other hand, has no such guarantee and could easily be rounded down. This is just my personal opinion, perhaps the authors had a different rationale.  
Either way, their approach should translate to java pretty effectively as the code is already "generic" and not tailored to a specific architecture or platform.

**Comment by woheller69** [comment]
Are there similar tweaks for things like

sin(11Pi) or sin(2000Pi) ? (See above)

For exact multiples of PI I fixed it already in my fork. But maybe there are better solutions to improve precision

Fo

**Comment by supsm** [comment]
I don't believe so. Neither gcc nor clang fixes this issue (they give the same result as java), and in fact the input value is simply not representable very well using double-precision floating point. [The closest representable value is 6283.18530718](https://float.exposed/0x40b88b2f704a940a), and [its sin value is as reported above](https://www.wolframalpha.com/input?i=sin%286283.18530717958674359%29).

**Comment by woheller69** [comment]
Maybe this library is an option?

https://github.com/eobermuhlner/big-math

**Comment by woheller69** [comment]
But probably that does not help either, as the input value does not come as BigDecimal. The whole Arity Engine would have to be changed...

**Comment by woheller69** [comment]
@supsm  I added your modifications and some test cases to my fork of Arity engine:
https://github.com/woheller69/ArityEngine

Here is an updated version of Arity. tanh and atanh (2%^9) now yield 5.12E-16

-uninstall
-remove .zip and install

[app-release.apk.zip](https://github.com/woheller69/Arity/files/12905442/app-release.apk.zip)

**Comment by woheller69** [comment]
This should fix some more issues with sin and cos calculations for multiples of PI or PI/2
https://github.com/woheller69/ArityEngine/commit/fab02d62e38fe5af723891b6c7d273cc1b7efca7

[app-release.apk.zip](https://github.com/woheller69/Arity/files/12907636/app-release.apk.zip)

**Comment by Murmele** [comment]
I found a similar problem:

4÷3×π×(((6371+0.1)×1000)^3−(6371×1000)^3×2÷1000. Dividing this formula by 1000:

4÷3×π×(((6371+0.1)×1000)^3−(6371×1000)^3×2÷1000÷1000 does not result in a lower exponent, only the mantissa changes

**Comment by woheller69** [comment]
If you want to divide the whole formula by 1000 you should put it in brackets

(4÷3×π×(((6371+0.1)×1000)^3−(6371×1000)^3×2÷1000)÷1000

**Comment by Murmele** [comment]
Sorry I completely overseen that one bracket is not closed
